server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -402,13 +402,14 @@ private Releasable acquireStore(Store store) {
         store.incRef();
         return Releasables.releaseOnce(() -> {
             final PlainActionFuture<Void> future = new PlainActionFuture<>();
-            threadPool.generic().execute(new ActionRunnable<>(future) {
-                @Override
-                protected void doRun() {
-                    store.decRef();
-                    listener.onResponse(null);
-                }
-            });
+            assert threadPool.generic().isShutdown() == false;
+            // TODO: We shouldn't use the generic thread pool here as we already execute this from the generic pool.
```

Comment:
@henningandersen Maybe we need to revisit the original solution (in https://github.com/elastic/elasticsearch/pull/45409) to avoid forking here.

Reply:
Yes, I agree to that. I think the best path is to make closing the resources async (except the one in the outer exception handler) such that we notify the recovery-response listener after completion of closing resources (we sometimes notify first and then close).